tools/build_langserver/langserver/handler.go

```diff
@@ -0,0 +1,152 @@
+	if request.Method != "initialize" && h.init == nil {
+		return nil, fmt.Errorf("server must be initialized")
+	}
+	h.conn = conn
+
+	methods := map[string]func(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error){
+		"initialize":      h.handleInit,
+		"initialzed":      h.handleInitialized,
+		"shutdown":        h.handleShutDown,
+		"exit":            h.handleExit,
+		"$/cancelRequest": h.handleCancel,
+	}
+
+	return methods[request.Method](ctx, request)
+
+}
+
+func (h *LsHandler) handleInit(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error) {
+	if h.init != nil {
+		return nil, errors.New("language server is already initialized")
+	}
+	if request.Params == nil {
+		return nil, &jsonrpc2.Error{Code: jsonrpc2.CodeInvalidParams}
+	}
+
+	var params lsp.InitializeParams
+	if err := json.Unmarshal(*request.Params, &params); err != nil {
+		return nil, err
+	}
+
+	// Set the Init state of the handler
+	h.mu.Lock()
+	core.FindRepoRoot()
+
+	h.RepoRoot = core.RepoRoot
+	h.SupportedCompletions = params.Capabilities.TextDocument.Completion.CompletionItemKind.ValueSet
+	params.EnsureRoot()
+	h.init = &params
+
+	h.mu.Unlock()
+
+	// Fill in the response results
+	TDsync := lsp.SyncIncremental
+	completeOps := &lsp.CompletionOptions{
+		ResolveProvider:   true,
+		TriggerCharacters: []string{"."},
+	}
+
+	sigHelpOps := &lsp.SignatureHelpOptions{
+		TriggerCharacters: []string{"{", ","},
+	}
+
+	log.Info("Initialize plz build file language server...")
+	return lsp.InitializeResult{
+		Capabilities: lsp.ServerCapabilities{
+			TextDocumentSync:           &TDsync,
+			HoverProvider:              true,
+			CompletionProvider:         completeOps,
+			SignatureHelpProvider:      sigHelpOps,
+			DefinitionProvider:         true,
+			TypeDefinitionProvider:     true,
+			ImplementationProvider:     true,
+			ReferenceProvider:          true,
+			DocumentFormattingProvider: true,
+			DocumentHighlightProvider:  true,
+			DocumentSymbolProvider:     true,
+		},
+	}, nil
+}
+
+func (h *LsHandler) handleInitialized(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error) {
+	if h.init != nil {
+		return nil, nil
+	}
+	// TODO(bnmetrics): Rethink!
+	return nil, nil
+}
+
+func (h *LsHandler) handleShutDown(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error) {
+	h.mu.Lock()
+	if h.IsServerDown {
+		log.Warning("Server is already down!")
+	}
+	h.IsServerDown = true
+	h.mu.Lock()
+	return nil, nil
+}
+
+func (h *LsHandler) handleExit(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error) {
+	h.handleShutDown(ctx, request)
+	h.conn.Close()
+	return nil, nil
+}
+
+func (h *LsHandler) handleCancel(ctx context.Context, request *jsonrpc2.Request) (result interface{}, err error) {
+	// TODO(bnmetrics): rethink this, I'm probably doing this wrong lol
```

Comment:
Yeah it looks like it will need more complexity - like the server will need to keep a map of request id -> context cancel function, and this function would cancel the appropriate one.

Tbh sounds like a classic thing we can put off for later :)